# Week 02 Assignment glucose level data


Missing data occurs commonly in many data applications. Especially wearable sensor devices encounter measurement errors resulting in missing data. Although pandas can handle missing data, for instance, to exclude missing data in the descriptive analysis, we might decide to reshape our data to improve quality. We might consider excluding certain rows, columns, or decide to impute our data if we take an argumentative approach. This week we will work with sensor data of a semester 01 project. The data is timeseries related but a lot of sensor errors occurred. First, we inspect the quality data. Then we will work with imputation, interpolation, and smoothing techniques to overcome this issue of the error readings. Finaly we plot the data. Visualization is a method to translate data into information. As a data scientist you should be able to choose an appropiate visualizations. 


Keywords: data loading, data inspection, data exploration, data cleaning, impute data, missing data, timeseries, pandas, visualization, interactive plots, overflow/underflow error

More to read
- https://fennaf.gitbook.io/bfvm22prog1/
- https://towardsdatascience.com/how-to-analyze-blood-glucose-data-with-python-data-science-packages-4f160f9564be


More about pandas and bokeh programming
- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.interpolate.html
- https://rubberduckdebugging.com/
- https://docs.bokeh.org/en/latest/docs/gallery.html


Learning objectives

- Further practice numpy / pandas
- Read, inspect, clean, and impute JSON data. Handle missing data
- Apply different interpolation/smoothing techniques
- Recognize overflow/underflow error
- Perform visual and statistical analysis for time series data
- Maintain development environment 
- Apply coding standards and FAIR principles


Note: you can bring your own sensor data, but for the practical you need some missing data points. 

By the end of this week the student can:

- load a json dataset 
- typecast the Pandas DataFrame to appropiate data types
- inspect the dataset for quality and metadata information
- add a column with interpolated data in Pandas DataFrame
- perform visual analysis

please add your own topics you want to learn here: https://padlet.com/ffeenstra1/69vbvy0l8t32rq55


The assignment consists of 6 parts:

- [part 1: load the data](#0)
     - [Exercise 1.1](#ex-11)
- [part 2: prepare for inspection](#1)
     - [Exercise 2.1](#ex-21)
- [part 3: inspect the data](#2)
     - [Exercise 3.1](#ex-31)
- [part 4: interpolate the data](#3)
     - [Exercise 4.1](#ex-41)
- [part 5: visualize the data](#4)
     - [Exercise 5.1](#ex-51)
- [part 6: Challenge](#5)
     - [Exercise 6.1](#ex-61)

Part 1 and 5 are mandatory, part 6 is optional (bonus)
Mind you that you cannot copy code without referencing the code. If you copy code you need to be able to explain your code verbally and you will not get the full score. 


<a name='0'></a>
## Part 1: Load the data

Instructions: Load the json datafile `glucose.json` into a pandas dataframe. Check your dataframe with a `.head()` to compare with the expected outcome. 

Preferably we read the data not with a hard coded data path but using a config file. See https://fennaf.gitbook.io/bfvm22prog1/data-processing/configuration-files/yaml

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>json.load() method reads a file, pd.read_json converts it to a Pandas DataFrame</li>
    <li>when loading into a Pandas DataFrame use records orientation </li>
</ul>
</details>

<a name='ex-11'></a>
### Code your solution

In [372]:
import yaml
with open("config.yaml", 'r') as stream:
    config = yaml.safe_load(stream)
file = config['data_glucose']

In [373]:
file

'../study_cases/data/glucose.json'

In [374]:
# CODE YOUR SOLUTION HERE
import pandas as pd
import json
import numpy as np
f = open(file)
data = json.load(f)

df = pd.read_json(data, orient='records')
df.head()

ID              time  recordtype glucose
0  2845.0  2019-04-25 00:08           1     109
1  2850.0  2019-04-25 00:50           1        
2  2877.0  2019-04-25 07:02           1     123
3  2881.0  2019-04-25 07:34           1     158
4  2886.0  2019-04-25 08:19           1

#### Expected outcome: 

<a name='1'></a>
## Part 2: Prepare the data

Check the datatypes of your dataframe. The `glucose` field should be an integer, the `time` field should have a datetime format. If the datatypes are different you should typecast them to the right format.
Make sure that your dataset is sorted by the time column


<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>use astype() method or pandas.DataFrame.to_datetime() for instance</li>
    <li>make sure that the empty spaces are filled with NaN. Use errors='coerce'</li>
    <li>set_index(), sort_index() and reset_index() are helpful to sort on index</li>
</ul>
</details>

<a name='ex-21'></a>
### Code your solution

In [375]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136 entries, 0 to 135
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   ID          136 non-null    float64
 1   time        136 non-null    object 
 2   recordtype  136 non-null    int64  
 3   glucose     54 non-null     object 
dtypes: float64(1), int64(1), object(2)
memory usage: 4.4+ KB


In [376]:
# Set ID to int
df['ID'] = df['ID'].astype(int)
# Set time to datetime
df['time'] = pd.to_datetime(df['time'],format='%Y-%m-%d %H:%M')

# Change the datatype of glucose to integers
df['glucose'] = pd.to_numeric(df['glucose'], errors='coerce')


In [377]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 136 entries, 0 to 135
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   ID          136 non-null    int64         
 1   time        136 non-null    datetime64[ns]
 2   recordtype  136 non-null    int64         
 3   glucose     52 non-null     float64       
dtypes: datetime64[ns](1), float64(1), int64(2)
memory usage: 4.4 KB


#### Expected outcome: 

In [378]:

# Set and sort the index by time column
df.set_index(['time'], inplace = True)
df.sort_index(inplace=True)
df.reset_index(inplace = True)
df.head()

time                   ID  recordtype  glucose
0 2019-04-25 00:08:00                 2845           1    109.0
1 2019-04-25 00:14:00 -9223372036854775808           0      NaN
2 2019-04-25 00:29:00 -9223372036854775808           0      NaN
3 2019-04-25 00:44:00 -9223372036854775808           0      NaN
4 2019-04-25 00:50:00                 2850           1      NaN

In [379]:
df

time                   ID  recordtype  glucose
0   2019-04-25 00:08:00                 2845           1    109.0
1   2019-04-25 00:14:00 -9223372036854775808           0      NaN
2   2019-04-25 00:29:00 -9223372036854775808           0      NaN
3   2019-04-25 00:44:00 -9223372036854775808           0      NaN
4   2019-04-25 00:50:00                 2850           1      NaN
..                  ...                  ...         ...      ...
131 2019-04-25 23:02:00 -9223372036854775808           0      NaN
132 2019-04-25 23:18:00 -9223372036854775808           0      NaN
133 2019-04-25 23:31:00                 3062           1    111.0
134 2019-04-25 23:33:00 -9223372036854775808           0      NaN
135 2019-04-25 23:48:00 -9223372036854775808           0      NaN

[136 rows x 4 columns]

<a name='2'></a>
## Part 3: Inspect the data

Now that we prepared the data we are going to inspect the data to get more familiar with the data. You can answer the questions below but feel free to explore the data further by plotting and or statistics, group by overviews and so on. 

To be answered at least:
- what is the quantity of the data (observations and features)
- what is the percentage missing data for glucose?
- is there a specific pattern for glucose value and recordtype?
- what is the time range?

Code the solutions to your answers. Create meaningful overviews or statistics

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>In the week 01 assignment some functions were explained to inspect missing values</li>
    <li>In the week 01 assignment some functions were explained to groupby value</li>
</ul>
</details>

<a name='ex-31'></a>
### Code your solution

In [380]:
print("The dataframe has", len(df), "observations")
print("Total amount of ID's are",len(df.groupby('ID')))
print("Total number of missing values is", df['glucose'].isna().sum())
#print("Total number of missing values is", df.isnull().sum().sum())
#print("Total amount of time records are",len(df.groupby('time')))

# Specific pattern for glucose value and recordtype
print("Glucose values per", df.groupby('recordtype')['glucose'].count())

The dataframe has 136 observations
Total amount of ID's are 55
Total number of missing values is 84
Glucose values per recordtype
0     0
1    52
Name: glucose, dtype: int64


In [381]:
df['time']

0     2019-04-25 00:08:00
1     2019-04-25 00:14:00
2     2019-04-25 00:29:00
3     2019-04-25 00:44:00
4     2019-04-25 00:50:00
              ...        
131   2019-04-25 23:02:00
132   2019-04-25 23:18:00
133   2019-04-25 23:31:00
134   2019-04-25 23:33:00
135   2019-04-25 23:48:00
Name: time, Length: 136, dtype: datetime64[ns]

In [382]:
#Time range
#df['time'] = pd.date_range("2019-04-25", periods=136 , freq='D')


In [383]:
# Percentage missing data for glucose
missing = df['glucose'].isna().sum() / len(df)
print("Percentage missing data is", missing)

Percentage missing data is 0.6176470588235294


#### Expected outcome percentage missing data
0.6176470588235294

<a name='3'></a>
## Part 4: Interpolate the data

A lot of data is missing. Use interpolation to fill the missing values. Create a new column with the interpolated data. Take an argumentative approach. Select an interpolation method that suits the nature of the data and explain your choice. Mind you that the expected outcome of the interpolation values can differ from the example below

<details>    
<summary>
    <font size="3" color="darkgreen"><b>Hints</b></font>
</summary>
<ul><li>use Pandas.DataFrame.interpolate() method</li>
</ul>
</details>

<a name='ex-41'></a>
### Code your solution

In [384]:
#CODE YOUR SOLUTION HERE
df['interpolated'] = df["glucose"].interpolate(method='linear', limit_diretion='forward')

df.head()

time                   ID  recordtype  glucose  interpolated
0 2019-04-25 00:08:00                 2845           1    109.0    109.000000
1 2019-04-25 00:14:00 -9223372036854775808           0      NaN    109.466667
2 2019-04-25 00:29:00 -9223372036854775808           0      NaN    109.933333
3 2019-04-25 00:44:00 -9223372036854775808           0      NaN    110.400000
4 2019-04-25 00:50:00                 2850           1      NaN    110.866667

#### Example outcome

<a name='4'></a>
## Part 5: Plot the data

Create a plot with the original data and the interpolated data. Consider what the best representation is for visualisation of actual values and modelled/imputed values. Provide argumentation for the plot of choice. 

Mind you, there are several types of plots. The most used are: 
- Compare values->	Bar, boxplot, violin plot, scatter, line. 
- Show composition-> Stacked bar, pie, mosaic, area, burtin.
- Show distribution-> 	Histogram, density, qqplot, bubble. 
- Analyzing trend-> Line, bar, dual axis 
- Relations-> Scatter, heatmap


### Code your solution

In [395]:
from bokeh.io import output_notebook
from bokeh.plotting import figure, show
from bokeh.layouts import gridplot
from bokeh.io import output_notebook
from bokeh.plotting import ColumnDataSource
from bokeh.models import Legend
from bokeh.models import DatetimeTickFormatter

output_notebook()

Loading BokehJS ...

In [386]:
# ID has minus values that shouldn't be there
# Select only those higher then 0
df = df[df.ID > 0]

In [406]:
p = figure(title='Glucose levels', x_axis_type="datetime", plot_width=1500)
p.xaxis.axis_label = 'Time'
p.yaxis.axis_label = 'Glucose level (mg/DL)'

source = ColumnDataSource(df)
g = p.circle(x='time', y='glucose', source=source, color='red', size=10, fill_color=None)
i = p.circle(x='time', y='interpolated', source=source, color='blue', size=4)
p.line(x='time', y='interpolated', source=source, color='lightgrey')


legend = Legend(items=[('glucose',[g]), ('interpolated',[i])], location='center')
p.add_layout(legend, 'right')

p.xaxis.formatter=DatetimeTickFormatter(days="%m/%d %H:%M",
months="%m/%d %H:%M",
hours="%m/%d %H:%M",
minutes="%m/%d %H:%M")

show(p)

<a name='6'></a>
## Part 6: Challenge

It might even be interesting to introduce a widget in which you can select different methods to interpolate.
1. Can you improve the interpolation by choosing an other method?
2. Can you add an rolling mean line? 
2. Can you improve the plot by making it interactive?

<a name='ex-61'></a>
### Code your solution

In [388]:
#CODE YOUR SOLUTION HERE